### ISSUE TITLE
Update to Mongoose v6 and MongoDB v4

### ISSUE THREAD

**Issue by Davies-Owen** [devconcept/multer-gridfs-storage#377]
First I want to thank you for this wonderful package. 

Do you know when this project will be updated to use Mongoose v6 and MongoDB v4? I see that PRs for these issues are already open: 
https://github.com/devconcept/multer-gridfs-storage/pull/375
https://github.com/devconcept/multer-gridfs-storage/pull/376

I have a project that relies on both `multer-gridfs-storage` and `mongoose`, and I can't update to mongoose v6 until this package is updated due to type conflicts.

**Describe your problem**

**Environment**

**Additional context**

**Comment by marlonklc** [comment]
> First I want to thank you for this wonderful package.
> 
> Do you know when this project will be updated to use Mongoose v6 and MongoDB v4? I see that PRs for these issues are already open:
> #375
> #376
> 
> I have a project that relies on both `multer-gridfs-storage` and `mongoose`, and I can't update to mongoose v6 until this package is updated due to type conflicts.
> 
> **Describe your problem**
> 
> **Environment**
> 
> **Additional context**

up

**Comment by devconcept** [comment]
Thank you for taking the time to report this

Maybe there is a change in mongoose v6 that I'm not aware of but this package uses a mongoose like connection, not an actual mongoose object, which means that unless the api has drastically changed it should be compatible. If you notice the mongoose dependency is a [dev dependency](https://github.com/devconcept/multer-gridfs-storage/blob/4839e481715891fe0b2631abd373300bf1c93669/package.json#L64) because I use mongoose on the test suite. The PR has no relation with your code.

As for mongodb the compatible versions stated on the package is v2 as you can see [here](https://github.com/devconcept/multer-gridfs-storage/blob/4839e481715891fe0b2631abd373300bf1c93669/package.json#L45) and I have reports that is currently working. The version you see is the one that get's installed when you want to contribute on the module, not in your production code.

The internal change to mongodb v4 is a big one so it will take some time to have a fix. This shouldn't be a problem for your code anyway.

Please give me feedback if my reasoning is not correct. Otherwise I'll be closing this issue on a week due to inactivity.

**Comment by Davies-Owen** [comment]
Thanks for the clarification about mongoose. As for MongoDB, let me explain my problem more clearly. I am creating the `GridFsStorage` object like this:
```
const storage = new GridFsStorage({
  db: <my already-created db>,
  file: (_req, file) => {
    this.logger.debug('Multer saving a file', file);
  },
});
```
Since the type of my already-created db comes from mongoose v6, which uses mongoDB v4 under the hood, TypeScript gives an error:
```
Type 'Db' is not assignable to type 'DbTypes | Promise<DbTypes>'.
  Type 'Db' is missing the following properties from type 'Db': serverConfig, bufferMaxEntries, native_parser, executeDbAdminCommand, and 16 more.
```
Essentially the problem I have is that mongoose v6 and `multer-gridfs-storage` are incompatible at the moment because they use different versions of mongodb. 
This would be fixed by updating mongo to v4 in this project, but I understand that that's a big undertaking.

**Comment by devconcept** [comment]
Can you please verify first that you don't have two versions of mongodb installed?

It might be the case that you have mongodb in your system specified with one version and mongoose internally uses another creating incompatibility.

If you check [the code](https://github.com/devconcept/multer-gridfs-storage/blob/4839e481715891fe0b2631abd373300bf1c93669/src/types/db-types.ts#L1) is a simple reference to a Db object. If the types are correct it should work.

**Comment by devconcept** [comment]
This issue will be closed in two days if no more feedback is received

**Comment by Davies-Owen** [comment]
I haven't had much time to investigate. I will look into this once I get more time and file a new issue if I still think there's a real problem here. Feel free to close the issue. Thanks for your help.

**Comment by bwobbones** [comment]
There is still a problem with the new mongo driver, it removes the `topology` object that `gridfs.js` file uses to gather the connection status:

https://github.com/devconcept/multer-gridfs-storage/blob/4839e481715891fe0b2631abd373300bf1c93669/src/gridfs.ts#L468

I now get an error from within `multer-gridfs-storage` as I try and instantiate `GridFsStorage`: `Cannot read property 'isConnected' of undefined`

Here is the (unhelpful) deprecation notice: https://github.com/mongodb/node-mongodb-native/blob/4.0/docs/CHANGES_4.0.0.md#removed-deprecations

The 3.7 API that has `isConnected`: http://mongodb.github.io/node-mongodb-native/3.7/api/Db.html
The 4.1 API where is has been removed: https://mongodb.github.io/node-mongodb-native/4.1/classes/Db.html

I think they are trying to say that if you are connected it doesn't help to check for the connection, so just deal with the error instead.  

Ideas for a PR?  I'm happy to contribute.  Maybe we could try using the connection and dealing with the error in `_updateConnectionStatus():`?

**Comment by devconcept** [comment]
I see now. Is not that mongoose is incompatible, is that by not providing the _client_ parameter the module assumes version 2.2 of mongodb. This is an easy fix and dealing with the error is the wrong approach since the module deals with behavior from versions 2.2, 3 and 4 and also the lack of parameters provided by the consumer of the module will make it unreliable. I will provide a fix on the next few days.

**Comment by devconcept** [comment]
A fix has been released v5.0.2. Please confirm that it works to close this issue.

**Comment by bwobbones** [comment]
It works!  Thanks for the quick turnaround.
